In [1]:
from tkinter import *
import os

In [2]:
def getdict(path):
    book = {}
    for line in open(path+'BX-Books.csv',encoding='ISO-8859-1'):
        line = line.replace('"',"")
        (ISBN,booktitle)=line.split(';')[0:2]
        book[ISBN]=booktitle
    p = {}
    count = 0
    for line in open(path+'BX-Book-Ratings.csv',encoding='ISO-8859-1'):
        line = line.replace('"', "") 
        line = line.replace("\\","")        
        (user,ISBN,rating)=line.split(';')
        try:
            if float(rating) > 0.0:
                p.setdefault(user,{})
                p[user][book[ISBN]]=float(rating)
        except ValueError:
            count+=1
            print ( user +" "+ ISBN )
        except KeyError:
            count+=1
            print (user + " " + ISBN)
    return p 
    

In [26]:
#userdata = getdict(path = 'C:\\Users\\AATTANG\\Downloads\\BX-CSV-Dump\\')

In [6]:
#userdata

In [7]:
#print out the list of books 
books = []
IDs = [] 
for keys, values in userdata.items() :
    IDs.insert(0,keys)
    for key ,value in values.items():
        books.insert(0,key)
        #print(key)
    #print (values)
#print(books)

In [8]:
#print user IDs
#userdata.keys()

In [12]:
#print(IDs)

to see whether they are no duplicate IDs

In [9]:
len(IDs),  len(set(IDs))


(77805, 77805)

shows they are no duplicate users users

In [10]:
if len(books) > len(set(books)):
   pass 

In [11]:
def allUnique(x):
    seen = set()
    return not any(i in seen or seen.add(i) for i in x)

In [12]:
allUnique(books)

False

 the shows that they arent unique variables in the data

In [13]:
#print unique list of books
unique_books = set(books)

In [14]:
len(unique_books),len(books)

(135394, 382807)

They are 135394 books in the dataset 

Function to get random list of books 

In [15]:
import random
def getrandombooks(listofbooks):
    list_of_random_items =  random.sample(listofbooks,10)
    return list_of_random_items

In [16]:
results = getrandombooks(unique_books)
results

['The Quilt: Stories from the Names Project',
 'Winter Soldier (Soldiers Of Fortune) (Silhouette Desire, No 1351)',
 "The Interpreter (Ballantine Reader's Circle)",
 'Black Holes and Baby Universes and Other Essays',
 "Midnight's Lair",
 'G.A.S. (GAS). Die Trilogie der Stadtwerke.',
 "The Bodyguard's Story: Diana, the Crash, and the Sole Survivor",
 'Dominique Moceanu:: A Gymnastics Sensation',
 'The Seven Storey Mountain: Fiftieth-Anniversary Edition',
 'El Vizconde Demediado']

User based collaborative filtering and item based collaborative filtering

In [17]:
from random import randint
import tkinter as RG
from math import *

In [18]:
def euclidean_distance(p, person1, person2):
    number=0
    for item in p[person1]:
        if item in p[person2]:
            number += 1              
    if number == 0: 
        return 0
    sum_of_squares = sum([pow(p[person1][item]-p[person2][item],2) for item in p[person1] if item in p[person2]])

    return 1 / (1 + sum_of_squares)

def sim_pearson(p,p1,p2):
    sim = {}
    for item in p[p1]:
        if item in p[p2]: 
            sim[item] = 1
    if len(sim) == 0:
        return 0
    n = len(sim)
    sum1 = sum([p[p1][it] for it in sim])
    sum2 = sum([p[p2][it] for it in sim])
    sum1Sq = sum([pow(p[p1][it],2) for it in sim])
    sum2Sq = sum([pow(p[p2][it],2) for it in sim])
    pSum = sum([p[p1][it] * p[p2][it] for it in sim])
    num = pSum - (sum1 * sum2/n)
    den = sqrt((sum1Sq - pow(sum1,2)/n) * (sum2Sq - pow(sum2,2)/n))
    if den == 0:
        return 0
    r = num/den
    return r

In [19]:
#Function to transform Person,book to book, person in the dictionary
def transformdictionary(d):
    results = {}
    for person in d:
        for item in d[person]:
            results.setdefault(item,{})

            #Flip book and userID
            results[item][person] = d[person][item]
    return results

In [20]:
flippeddict = transformdictionary(userdata)

Looking at the distance between the each variable for the transformed dictionary using euclidean distance 

In [22]:
def topmatches(p,person,n=5,similarity=euclidean_distance):
    scores = [(similarity(p,person,other),other)
                for other in p if other != person]
    scores.sort()
    scores.reverse()
    return scores[0:n]


User Interface

In [27]:
def signup():
    window4 = RG.Toplevel(root)
    window4.geometry("300x150+75+75")
    booklist1=Label(window4,text = "name")
    value = StringVar()
    entryvalue1 = Entry(window4,textvariable = value)
    booklist1.grid(row=0, sticky=E)
    entryvalue1.grid(row=0, column=1)
    bookList2 = Label(window4,text= "email ")
    value1 = StringVar()
    entryvalue2 = Entry(window4,textvariable = value1)
    bookList2.grid(row=0, sticky=E)
    entryvalue2.grid(row=0, column=1)
    buttonSubmit= Button(window4, text="submit", command= lambda: getuser())
    buttonSubmit.grid(row= 5,column=1)
    
    def getuser():
        A =random.sample(IDs,1)
        window5 = RG.Toplevel(window4)
        window5.geometry("300x150+75+75")    
        var = StringVar()
        booklst= A    
        var.set(booklst)  
        labelName = Label(window5, text = "This is your user ID")        
        bookList = Label(window5, textvariable=var) 
        labelName.grid(row=0, sticky=E)
        bookList.grid(row=4)
        window5.geometry("350x250+75+75")
        
def putID():
    window = RG.Toplevel(root)
    window.geometry("300x150+75+75")
    bookList = Label(window, text="Enter ID ")  
    Variable = StringVar()
    entryValue = Entry(window, textvariable=Variable)
    bookList.grid(row=0, sticky=E)
    entryValue.grid(row=0, column=1)
    buttonSubmit= Button(window, text="OK", command= lambda: getRec(Variable))
    buttonSubmit.grid(row= 5,column=1)
# this uses user based         
    def getRec(i):
        similarity = euclidean_distance
        try:
            value = i.get()
        except ValueError:
            pass
        totals={}
        similarity_sum={}
        for otherperson in userdata:
            if otherperson==value: continue
            sim=similarity(userdata,value,otherperson)
            if sim<=0: continue
            for item in userdata[otherperson] :
                if item not in userdata[value] or userdata[value][item]==0:
                    totals.setdefault(item,0)
                    totals[item]+=sim*userdata[otherperson][item]
                    similarity_sum.setdefault(item,0)
                    similarity_sum[item]+=sim
        scores=[(totals[item]/similarity_sum[item],item) for item in totals.keys()]
        scores.sort()
        scores.reverse()   
        window1 = RG.Toplevel(window)
        window1.geometry("300x150+75+75")       
        var = StringVar()
        booklst= '\n'.join([x[1] for x in scores[0:8]])       
        var.set(booklst)       
        labelName = Label(window1, text = "Recommendations based on Similar User")        
        bookList = Label(window1, textvariable=var)   
        labelName.grid(row=0, sticky=E)
        bookList.grid(row=4)
        window1.geometry("350x250+75+75")
# this implements item based on random selection of a book
def randombooks():
    window2 = RG.Toplevel(root)
    window2.geometry("300x150+75+75")
    bookList1 = Label(window2, text="click 'ok' to start")
    bookList1.grid(row=0, sticky=E)
    buttonSubmit= Button(window2, text="OK", command= lambda: rando(books))
    buttonSubmit.grid(row= 5,column=1)
  
    def rando(result):
        items = random.sample(result,1)
        window3 = RG.Toplevel(window2)
        window3.geometry("300x150+75+75")
        var = StringVar()
        book= '\n'.join([x for x in items[0:1]])  
        var.set(book)
        labelName = Label(window3, text = "Start reading this")        
        bookList = Label(window3, textvariable=var)
        buttonSubmit1= Button(window3, text="get the next", command= lambda: simila(items))
        buttonSubmit1.grid(row= 5,column=1)
        labelName.grid(row=0, sticky=E)
        bookList.grid(row=4)
        window3.geometry("350x250+75+75")
        
        def simila(item):
            z =item[0]
            lol =topmatches(flippeddict,z,n=8,similarity=sim_pearson)
            window6 = RG.Toplevel(window3)
            window6.geometry("300x150+75+75")
            var = StringVar()
            book= '\n'.join([x[1] for x in lol])
            var.set(book)
            labelName1 = Label(window6, text = "if you liked that book you may like these")        
            bookList1 = Label(window6, textvariable=var)
            labelName1.grid(row=0, sticky=E)
            bookList1.grid(row=4)
            window6.geometry("350x250+75+75")

In [28]:
root = Tk()
root.geometry("300x200+200+200")
theLabel = Label(root, text="Welcome to Readers Guide")
theLabel.pack()
topFrame = Frame(root)
topFrame.pack(fill=BOTH, expand=True)
buttonsignup = Button(topFrame,text = 'sign in',command =signup )
buttonsignup.pack()
buttonBook = Button(topFrame, text="start reading(guest)",command =randombooks)
buttonBook.pack()
buttonRec = Button(topFrame, text="Book Recommendation",command=putID)
buttonRec.pack()

root.mainloop()

# Evaluation


In [192]:
def evaluation(user,k,df1):
    usersim = []
    simu = []
    users_products = list(np.nonzero(df1[user]))
    books = users_products[0][0]
    user_review = df1[user,books]
    users_with_prod = np.where(df1[:,books]!= 0)[0]
    for i in users_with_product:
        if i != user:
            sim = topmatches(df1,user,n=5,similarity=sim_pearson)
            usersim.append([sim,i])
    usersim.sort()
    usersim.reverse()
    for item in usersim[:k]:
        simu.append(item[1])
    rat = 0
    for kuser in simu:
        rat+=df1[kuser][books]
    pred_rating = rat/k
    return pred_rating,users_review
    

In [200]:
def cal_mae(df1,ratio,k):
    num_users = np.shape(df1)[0]
    list_of_users = list(range(0,num_users))
    np.random.shuffle(list_of_users)
    subset_size = int(num_users*ratio)
    print(subset_size)
    subset_users =list_of_users[0:subset_size]
    print(subset_users)
    abserror = 0
    count = 0
    for wuser in subset_users:
        pred,user = evaluation(wuser,k,df1)
        abserror += abs(pred - userr)
        count += 1
    MAE = abserror/count
    print("mean absolute error",MAE)